File: 2025/day03/Day3.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	total :=0
	for scanner.Scan() {
		line := scanner.Text()
		n := len(line)
		k := 12
		result := make([]byte, 0,20)
		start := 0
		for k > 0 {
			maxDigit := -1
			maxIndex := start
			end := n - k
			for i := start; i <= end; i++ {
				d := int(line[i] - '0')
				if d > maxDigit {
					maxDigit = d
					maxIndex = i
				}
			}
			result = append(result, byte(maxDigit)+'0')
			start = maxIndex + 1
			k--
		}
		val, err := strconv.Atoi(string(result))
		if err != nil {
			panic(err)
		}
		total += val
	}
	fmt.Println(total)
}

File: 2025/day03/Day3.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

int main(){
    char line[1024];
    long long total = 0;
    while (fgets(line, sizeof(line), stdin)) {
        line[strcspn(line, "\n")] = 0;

        int n = strlen(line);
        int k = 12;
        char result[20];
        int pos = 0;

        int start = 0;

        while (k > 0) {
            int max_digit = -1;
            int max_index = -1;

            int END = n - k;

            for (int i = start; i <= END; i++) {
                int d = line[i] - '0';
                if (d > max_digit) {
                    max_digit = d;
                    max_index = i;
                }
            }

            result[pos++] = line[max_index];
            start = max_index + 1;
            k--;
        }

        result[pos] = '\0';
        total = total + strtoll(result,NULL,10);
    }
    printf("%lld",total);
}

